priv/migrations/20211220163202_index_ga_attach_txs.ex

```diff
@@ -0,0 +1,126 @@
+defmodule AeMdw.Migrations.IndexGaAttachTxs do
+  @moduledoc """
+  Indexes ga_attach transactions in a way that transactions and calls
+  related to the smart contract of the GA might be searched respectively by
+  `/txs` and `/contract/calls` requests using the contract_id field.
+  """
+  alias AeMdw.Application, as: MdwApp
+  alias AeMdw.Node, as: AE
+
+  alias AeMdw.Db.Model
+  alias AeMdw.Db.Sync.Transaction
+  alias AeMdw.Log
+  alias AeMdw.Mnesia
+  alias AeMdw.Sync.Supervisor, as: SyncSup
+
+  require Model
+  require Ex2ms
+  require Logger
+
+  @fortuna_txi_begin 2129380
+
+  @doc """
+  Reindexes ga_attach_tx transactions and updates stats with contracts count.
+  """
+  @spec run(boolean()) :: {:ok, {non_neg_integer(), pos_integer()}} | {:error, :count_mismatch}
+  def run(from_startup?) do
+    begin = DateTime.utc_now()
+
+    if not from_startup? and :ok != Application.ensure_started(:ae_mdw) do
+      Log.info("Ensure sync tables...")
+      SyncSup.init_tables()
+      MdwApp.init_public(:contract_cache)
+      MdwApp.init_public(:db_state)
+    end
+
+    txs_found = lookup_txs()
+    indexed_count = reindex_txs(txs_found)
+
+    if indexed_count == length(txs_found) do
+      duration = DateTime.diff(DateTime.utc_now(), begin)
+      Log.info("Indexed #{indexed_count} records in #{duration}s")
+
+      {:ok, {indexed_count, duration}}
+    else
+      {:error, :count_mismatch}
+    end
+  end
+
+  defp lookup_txs() do
+    txi_spec =
+      Ex2ms.fun do
+        Model.type(index: {:ga_attach_tx, txi}) when txi >= @fortuna_txi_begin -> txi
+      end
+
+    {:atomic, txi_list} = :mnesia.transaction(fn ->
+      :mnesia.select(Model.Type, txi_spec, :read)
+    end)
+    Log.info("Found #{length(txi_list)} :ga_attach_tx(s)...")
+
+    {:atomic, bi_txs_list} = :mnesia.transaction(fn ->
+      Enum.map(txi_list, fn txi ->
+        [Model.tx(id: hash, block_index: bi)] = :mnesia.read(Model.Tx, txi, :read)
+        {bi, hash, txi}
+      end)
+    end)
+    Log.info("Found #{length(bi_txs_list)} :ga_attach_tx(s)...")
+
+    bi_txs_list
+  end
+
+  defp reindex_txs(bi_txs_list) do
+    bi_txs_list
+    |> Enum.group_by(fn {{height, _mbi}, _tx_hash, _txi} -> height end)
+    |> Enum.map(&sync_generation_gaattach_txs/1)
+    |> Enum.sum()
+  end
+
+  defp sync_generation_gaattach_txs({height, bi_txs_list}) do
+    txs_mutations =
+      height
+      |> AE.Db.get_micro_blocks()
+      |> Enum.with_index()
+      |> Enum.flat_map(fn {mblock, mbi} ->
+        has_tx_to_sync? = Enum.any?(bi_txs_list, fn {{_height, tx_mbi}, _tx_hash, _txi} ->
+          tx_mbi == mbi
+        end)
+
+        if has_tx_to_sync? do
```

Comment:
Why not use the functional Enum primitives here?
```elixir
|> Enum.filter(fn {mblock, mbi} -> Enum.any?(...) end)
|> Enum.flat_map(fn {mblock, mbi} ->
  # ...
end)
```

Reply:
Done. Could not retest the migration from scratch though because of the sync issue with higher priority.